pkg/signer/file.go

```diff
@@ -0,0 +1,61 @@
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package signer
+
+import (
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/rsa"
+	"fmt"
+
+	"github.com/sigstore/sigstore/pkg/signature"
+	"go.step.sm/crypto/pemutil"
+)
+
+// File returns a file-based signer and verifier, used for local testing
+type File struct {
+	crypto.Signer
+}
+
+func NewFileSigner(keyPath, keyPass string) (*File, error) {
+	opaqueKey, err := pemutil.Read(keyPath, pemutil.WithPassword([]byte(keyPass)))
+	if err != nil {
+		return nil, fmt.Errorf("file: provide a valid signer, %s is not valid: %w", keyPath, err)
+	}
+	// Cannot use signature.LoadSignerVerifier because the SignerVerifier interface does not extend crypto.Signer
+	switch pk := opaqueKey.(type) {
+	case *rsa.PrivateKey:
+		signer, err := signature.LoadRSAPKCS1v15SignerVerifier(pk, crypto.SHA256)
+		if err != nil {
+			return nil, err
+		}
+		return &File{signer}, nil
+	case *ecdsa.PrivateKey:
+		signer, err := signature.LoadECDSASignerVerifier(pk, crypto.SHA256)
+		if err != nil {
+			return nil, err
+		}
+		return &File{signer}, nil
+	case ed25519.PrivateKey:
```

Comment:
Just curious why this is not a pointer when the others are? If it's intentional, mayhaps a comment? If it's common knowledge and only I'm confused, then fine :)

Reply:
Maybe add a test case for this as well in file_test?

Reply:
`ParsePKCS8PrivateKey` "returns a *rsa.PrivateKey, a *ecdsa.PrivateKey, or a ed25519.PrivateKey". I'm not sure why the decision was made to make only ed25519 not a pointer, but it's permeated through golang crypto.

Added a test, and switched to generating keys in-memory!